### PR TITLE
#6722 CSP: provide infrastructure to add CSP-support to PrimeFaces-Selenium

### LINKED ISSUE #6722: CSP: provide infrastructure to add CSP-support to PrimeFaces-Selenium
see https://github.com/primefaces-extensions/primefaces-integration-tests/pull/78

### PR DESCRIPTION
**Pull request by christophs78**

### CODE REVIEW

File: src/main/resources/META-INF/resources/primefaces/core/core.csp.js [hunk 1]
```diff
@@ -63,7 +69,29 @@ if (!PrimeFaces.csp) {
                 };
 
                 $(element).off(jqEvent).on(jqEvent, jsWrapper);
+
+                //TODO: Depending on some (debug-)switch?
```
Comment (melloware) [inline src/main/resources/META-INF/resources/primefaces/core/core.csp.js]: Yes add `if (PrimeFaces.isDevelopmentProjectStage()) {` so this only runs when JSF stage is development.

File: src/main/resources/META-INF/resources/primefaces/core/core.csp.js [hunk 2]
```diff
@@ -63,7 +69,29 @@ if (!PrimeFaces.csp) {
...
+         */
+        getRegisteredEvent: function(id, event) {
+            if (this.registeredEvents.has(id)) {
+                var shortenedEvent = event.substring(2, event.length),
+                    jqEvent = shortenedEvent + '.' + id;
+
+                return this.registeredEvents.get(id).get(jqEvent);
             }
+            return "";
         },
```
Comment (melloware) [inline src/main/resources/META-INF/resources/primefaces/core/core.csp.js:103]: what about another helper than returns a boolean like `hasRegisteredEvent` which we can use in Selenium to determine whether this should be AJAX or not?

**Comment by tandraschko** [comment]
Im a bit unsure about this topic. Will think about ir

File: src/main/resources/META-INF/resources/primefaces/core/core.csp.js (hunk 1 above)
Comment (christophs78) [reply src/main/resources/META-INF/resources/primefaces/core/core.csp.js]: Are we sure integration-tests based on PrimeFaces Selenium are only executed in JSF development-stage?

File: src/main/resources/META-INF/resources/primefaces/core/core.csp.js (hunk 2 above)
Comment (christophs78) [reply src/main/resources/META-INF/resources/primefaces/core/core.csp.js:103]: good idea - done

**Comment by christophs78** [comment]
What do you think about replacing
```javascript
/**
 * Map containing registered event listeners per element.
 * @type {Map<string, Map<string, string>>}
 */
registeredEvents: new Map(),
```
with
```javascript
/**
 * Map containing registered event listeners per element.
 * @type {Map<string, Map<string, boolean>>}
 */
hasRegisteredEvents: new Map(),
```
This would reduce our memory-footprint to an absolute minimum.

**Comment by tandraschko** [comment]
Sounds better but PFSelenium needs to check the script content (isAjaxScript)?!

**Review by melloware — APPROVED** [review]

**Comment by melloware** [comment]
@tandraschko He is doing that in this PrimeSelenium PR: https://github.com/primefaces-extensions/primefaces-selenium/pull/63

**Comment by tandraschko** [comment]
i dont get how its working if you just have the info "there is a onclick script available" - you also need to search trough the event-script-string actually?

**Comment by christophs78** [comment]
Do we add non-ajax scripts via `PrimeFaces.csp.register`?
When yes we may add some logic like to following to core.csp.js:
`return script.contains("PrimeFaces.ab(") || script.contains("pf.ab(") || script.contains("mojarra.ab(") || script.contains("jsf.ajax.request");`
That´s the logic currently placed in `org.primefaces.extensions.selenium.component.base.ComponentUtils#isAjaxScript`.

**Comment by christophs78** [comment]
Or do we prefer to store the whole script inside the map like we had before?

**Comment by tandraschko** [comment]
AFAIR we do it for every script 
We could do it but -1 for production

**Comment by christophs78** [comment]
When we do it only for JSF dev-stage this means, you must turn your application into dev-stage to use PrimeFaces Selenium together with CSP. We would need to add this to the documentation.

Our own integration tests currently run in production-stage.
```xml
<context-param>
    <param-name>javax.faces.PROJECT_STAGE</param-name>
    <param-value>Production</param-value>
</context-param>
```

To be honest i´m currently not sure what the right trade-offs are.
I understand it´s not ideal to store CSP-JS in a map for production stage.

**Comment by christophs78** [comment]
Updated this PR to track only ajaxified events inside our csp-map.

**Comment by melloware** [comment]
Personally I would do the `Primefaces.isDevelopmentStage()` because in my experience Selenium test sould not be run against a Production system. If you don't know you have a bug before it gets to PRD that is not the time to run Selenium.  Just my two cents.

**Comment by tandraschko** [comment]
I think they should run in stage systemtest

**Comment by melloware** [comment]
OK so we should add new helpers for `PrimeFaces.isNotProductionStage()`

**Comment by christophs78** [comment]
There may be stages like QS or integration which are not covered by JSF stage. In my organization we use JSF development stage only during local development.
SystemTest makes sense. I‘ll do this and we add this to PF Selenium Docu.

**Review by melloware — APPROVED** [review]

**Comment by tandraschko** [comment]
Another Idea: why not overwrite the method in PFSelenium?

**Comment by christophs78** [comment]
> Another Idea: why not overwrite the method in PFSelenium?

Good idea! I tried this and currently tun into a timing issue. We have to overwrite PrimeFaces.csp.register after PF-core.js is loaded but before this function is called the first time.
I tried this as part of PF Selenium onload.js. Looks like this comes to late. Some idea´s how to fix this?

```javascript
console.log("check PrimeFaces.csp...");
if (PrimeFaces.csp) {
    console.log("PrimeFaces.csp is active");

    PrimeFaces.csp.REGISTERED_AJAXIFIED_EVENT_LISTENERS = new Map();

    var cspRegisterOrig = PrimeFaces.csp.register;
    PrimeFaces.csp.register = function(id, event, js){
        console.log("PF Selenium - PrimeFaces.csp.register");
        debugger;

        if (event) {
            cspRegisterOrig(id, event, js);

            //Collect some basic information about registered ajaxified event listeners
            if (!PrimeFaces.csp.REGISTERED_AJAXIFIED_EVENT_LISTENERS.has(id)) {
                PrimeFaces.csp.REGISTERED_AJAXIFIED_EVENT_LISTENERS.set(id, new Map())
            }

            var jsAsString = js.toString();
            var ajaxified = (jsAsString.indexOf("PrimeFaces.ab(") >= 0) || (jsAsString.indexOf("pf.ab(") >= 0) || (jsAsString.indexOf("mojarra.ab(") >= 0) || (jsAsString.indexOf("jsf.ajax.request") >= 0);
            PrimeFaces.csp.REGISTERED_AJAXIFIED_EVENT_LISTENERS.get(id).set(jqEvent, ajaxified);
        }
    };

    PrimeFaces.csp.hasRegisteredAjaxifiedEvent = function(id, event) {
        console.log("PF Selenium - PrimeFaces.csp.hasRegisteredAjaxifiedEvent");
        debugger;

        if (PrimeFaces.csp.REGISTERED_AJAXIFIED_EVENT_LISTENERS.has(id)) {
            var shortenedEvent = event.substring(2, event.length),
                jqEvent = shortenedEvent + '.' + id;

            return PrimeFaces.csp.REGISTERED_AJAXIFIED_EVENT_LISTENERS.get(id).get(jqEvent);
        }
        return false;
    };
}
```

Console-debug.output:
grafik

**Comment by christophs78** [comment]
Please have a look at https://github.com/primefaces-extensions/primefaces-selenium/pull/63/files.
Specially at `PrimefacesSeleniumPhaseListener `and `pfselenium.core.csp.js`.

There is only one issue left:
`org.primefaces.csp.CspPartialResponseWriter` renders `sb.append("PrimeFaces.csp.register('");`

`pfselenium.core.csp.js` does following check:
```javascript
var ajaxified = (jsAsString.indexOf("PrimeFaces.ab(") >= 0) || (jsAsString.indexOf("pf.ab(") >= 0) || (jsAsString.indexOf("mojarra.ab(") >= 0) || (jsAsString.indexOf("jsf.ajax.request") >= 0);
```

But `org.primefaces.application.resource.MoveScriptsToBottomResponseWriter#mergeAndMinimizeInlineScripts` does `.replace("PrimeFaces.ab", "pf.ab")`.

So PF sends following JS to the client:
```javascript
var ajaxified = (jsAsString.indexOf("pf.ab(") >= 0) || (jsAsString.indexOf("pf.ab(") >= 0) || (jsAsString.indexOf("mojarra.ab(") >= 0) || (jsAsString.indexOf("jsf.ajax.request") >= 0);
```

And this check does not detect `PrimeFaces.csp.register` written by `org.primefaces.csp.CspPartialResponseWriter`.

We are so close to our target.

**Comment by tandraschko** [comment]
Hmm, i dont understand the "pf.ab" vs "PrimeFaces.ab" problem. Only when MOVE_SCRIPTS_TO_BOTTOM is active, its replaced and then PrimeFaces.ab will never occur?

**Comment by christophs78** [comment]
Integration tests currently set
```xml
<context-param>
	<param-name>primefaces.MOVE_SCRIPTS_TO_BOTTOM</param-name>
	<param-value>true</param-value>
</context-param>
```

Our trick should work independent of `primefaces.MOVE_SCRIPTS_TO_BOTTOM`.

Tried it with primefaces.MOVE_SCRIPTS_TO_BOTTOM=false....

This bring´s back our timing-issue:
grafik

grafik
gets executed before
grafik

Maybe because $(function is delayed by jQuery or Browser...

**Comment by christophs78** [comment]
Yep `$(function() { ... });` is a shortcut for `$(document).ready(function() { ... });`
https://stackoverflow.com/questions/7642442/what-does-function-do

**Comment by melloware** [comment]
Yep PF wraps in `$(function() { ... });` when not using Move to Bottom so the scripts are executed after the DOM has finished loading.  With MOVE_TO_BOTTOM the scripts all all moved to the bottom therefore they don't need the function wrapper because by definition they are running after the DOM has loaded.

I had to make this fix a while ago:
https://github.com/primefaces/primefaces/commit/fdfbb292f399968c884e5a9724ae02521203fc2f

To fix: https://github.com/primefaces/primefaces/issues/5455

**Comment by christophs78** [comment]
What a hell of a ride.
Finally got this working without modifiying PrimeFaces. So - from my perspective - we can close this PR.
Had been able to do this in PrimeFaces Selenium. 
Please look at https://github.com/primefaces-extensions/primefaces-selenium/pull/63/files.
I introduced `PrimefacesSeleniumPhaseListener` which adds `PrimeFacesSeleniumSystemEventListener` as `PreRenderViewEvent`. 
`PrimeFacesSeleniumSystemEventListener` adds `PrimeFacesSeleniumDummyComponent` to the component-tree.
And finally `PrimeFacesSeleniumDummyComponent` adds `pfselenium.core.csp.js` after `core.js` was added by PrimeFaces itself.
This works independent of JSF-stage, MOVE_TO_BOTTOM , ...

**Comment by melloware** [comment]
Excellent work I will close this PR and issue!   great job on fixing it all on the PF selenium side. I left you code review comments.

**Pull request closed without merging**